src/caf/toolkit/cost_utils.py

```diff
@@ -3,25 +3,422 @@
+
+    @property
+    def avg_vals(self) -> np.ndarray:
+        """Average values for each of the cost distribution bins."""
+        return self.df[self.avg_col].values
+
+    @property
+    def trip_vals(self) -> np.ndarray:
+        """Trip values for each of the cost distribution bins."""
+        return self.df[self.trips_col].values
+
+    @property
+    def band_share_vals(self) -> np.ndarray:
+        """Band share values for each of the cost distribution bins."""
+        trip_vals = self.trip_vals
+        return trip_vals / np.sum(trip_vals)
+
+    @staticmethod
+    def from_data(
+        matrix: np.ndarray,
+        cost_matrix: np.ndarray,
+        min_bounds: Optional[list[float] | np.ndarray] = None,
+        max_bounds: Optional[list[float] | np.ndarray] = None,
+        bin_edges: Optional[list[float] | np.ndarray] = None,
+    ) -> CostDistribution:
+        """Convert values and a cost matrix into a CostDistribution.
+
+        Parameters
+        ----------
+        matrix:
+            The matrix to calculate the cost distribution for. This matrix
+            should be the same shape as cost_matrix
+
+        cost_matrix:
+            A matrix of cost relating to matrix. This matrix
+            should be the same shape as matrix
+
+        min_bounds:
+            A list of minimum bounds for each edge of a distribution band.
+            Corresponds to max_bounds.
+
+        max_bounds:
+            A list of maximum bounds for each edge of a distribution band.
+            Corresponds to min_bounds.
+
+        bin_edges:
+            Defines a monotonically increasing array of bin edges, including the
+            rightmost edge, allowing for non-uniform bin widths. This argument
+            is passed straight into `numpy.histogram`
+
+        Returns
+        -------
+        cost_distribution:
+            An instance of CostDistribution containing the given data.
+
+        See Also
+        --------
+        `cost_distribution`
+        """
+        # Define columns names
+        min_col = "min"
+        max_col = "max"
+        avg_col = "ave"
+        trips_col = "trips"
+
+        # Calculate the cost distribution
+        bin_edges = _validate_bin_edges(
+            bin_edges=bin_edges,
+            min_bounds=min_bounds,
+            max_bounds=max_bounds,
+        )
+        distribution = cost_distribution(
+            matrix=matrix, cost_matrix=cost_matrix, bin_edges=bin_edges
+        )
+
+        # Covert data into instance of this class
+        df = pd.DataFrame(
+            {
+                min_col: bin_edges[:-1],
+                max_col: bin_edges[1:],
+                avg_col: (np.array(bin_edges[:-1]) + np.array(bin_edges[1:])) / 2,
+                trips_col: distribution,
+            }
+        )
+
+        return CostDistribution(
+            df=df,
+            min_col=min_col,
+            max_col=max_col,
+            avg_col=avg_col,
+            trips_col=trips_col,
+        )
```

Comment:
If this was a class method

```suggestion
        # Covert data into instance of this class
        df = pd.DataFrame(
            {
                cls.min_col: bin_edges[:-1],
                cls.max_col: bin_edges[1:],
                cls.avg_col: (np.array(bin_edges[:-1]) + np.array(bin_edges[1:])) / 2,
                cls.trips_col: distribution,
            }
        )

        return CostDistribution(df=df)
```

Reply:
Solved (alongside a bug or two it created) here: https://github.com/Transport-for-the-North/caf.toolkit/pull/27/commits/473c414343f34e1b61fa21f4b4926d5b7e85014b